### PR TITLE
[#102] Implement first version of co-log-polysemy

### LINKED ISSUE #102: Implement `co-log-polysemy` package
Just a proof of concept. `co-log` is based on the MTL approach. But to show that the approach is extensible, we can implement one more package based on free monads.

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #102

### CODE REVIEW

File: co-log-polysemy/src/Colog/Polysemy/Effect.hs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+This function allows to use extensible effects provided by the @polysemy@
+library with logging provided by @co-log@. You can construct 'LogAction'
+independently and then just pass to this function to tell how to log messages.
+
+Several examples:
+
+1. @'runLogAction' mempty@: interprets the 'Log' effect by ignoring all messages.
+2. @'runLogAction' 'Colog.Core.IO.logStringStdout'@: interprets 'Log' effect by
+   allowing to log 'String'.
```
Comment (vrom911) [inline co-log-polysemy/src/Colog/Polysemy/Effect.hs]:
```suggestion
   allowing to log 'String' to @stdout@.
```

File: co-log-polysemy/co-log-polysemy.cabal [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+license-file:        LICENSE
+author:              Kowainik
+maintainer:          xrom.xkov@gmail.com
+copyright:           2019 Kowainik
+category:            Logging, Contravariant, Comonad, Effects
+build-type:          Simple
+stability:           provisional
+extra-doc-files:     CHANGELOG.md
+                   , README.md
+tested-with:         GHC == 8.2.2
```
Comment (isovector) [inline co-log-polysemy/co-log-polysemy.cabal]: Really? I'm like 95% sure polysemy won't compile below 8.6 --- not for any fundamental reasons other than it's using some unnecessary `QuantifiedConstraints` right now.

File: co-log-polysemy/src/Colog/Polysemy/Effect.hs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+2. @'runLogAction' 'Colog.Core.IO.logStringStdout'@: interprets 'Log' effect by
+   allowing to log 'String' to @stdout@.
+-}
+runLogAction
+    :: forall m msg r a .
+       Member (Lift m) r
+    => LogAction m msg
+    -> Sem (Log msg ': r) a
+    -> Sem r a
+runLogAction (LogAction action) = interpret $ \case
```
Comment (isovector) [inline co-log-polysemy/src/Colog/Polysemy/Effect.hs]: You probably should add `INLINE` pragmas for all of your interpretations; failure to do so will miss a lot of performance optimizations in `polysemy`!

File: co-log-polysemy/co-log-polysemy.cabal (hunk 2 above)
Comment (chshersh) [reply co-log-polysemy/co-log-polysemy.cabal]: I wrote these lines before I tested on the CI and noticed that `polysemy` doesn't build on GHC < 8.6. That's why I have the following lines in the `.cabal` file:

```
  if impl(ghc < 8.6)
      buildable:       False
```

Just forgot to update metadata. Thanks for noticing this!

File: co-log-polysemy/src/Colog/Polysemy/Effect.hs (hunk 3 above)
Comment (chshersh) [reply co-log-polysemy/src/Colog/Polysemy/Effect.hs]: Thanks for the advice! Will do that :+1:

**Comment by chshersh** [comment]
@isovector Many thanks for your review! Please, feel free to mention this package in your documentation, it's a pleasure to me :blush:

File: co-log-polysemy/co-log-polysemy.cabal (hunk 2 above)
Comment (isovector) [reply co-log-polysemy/co-log-polysemy.cabal]: https://github.com/isovector/polysemy/issues/29 is the ticket for this work on `polysemy` --- I'm traveling for a few weeks but can tackle it when I'm home. Should't be more than a ten minute change!

**Pull request merged**